### ISSUE TITLE
Rollup ^3.0.0 cjs re-exports edge case with externalLiveBindings: false

### ISSUE THREAD

**Issue by yyx990803** [nodejs/cjs-module-lexer#79]
This is a new version of the same problem reported back in https://github.com/nodejs/cjs-module-lexer/issues/38

Given code

```js
export * from 'foo'
```

Default `cjs` output from Rollup (working, correctly detects re-exports from `foo`):

```js
var foo = require('foo');

Object.keys(foo).forEach(function (k) {
	if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
		enumerable: true,
		get: function () { return foo[k]; }
	});
});
```

With [`output.externalLiveBindings: false`](https://rollupjs.org/guide/en/#outputexternallivebindings) (not working, no re-exports detected):

```js
var foo = require('foo');

for (var k in foo) {
	if (k !== 'default' && !exports.hasOwnProperty(k)) exports[k] = foo[k];
}
```

Note: this was working previously in Rollup 2.x where the code generated was:

```js
var foo = require('foo');

Object.keys(foo).forEach(function (k) {
	if (k !== 'default' && !exports.hasOwnProperty(k)) exports[k] = foo[k];
});
```

[Reproduction](https://github.com/yyx990803/cjs-module-lexer-rollup-reexports)

- Install deps
- `npm run build` (which builds `src/index.js` with Rollup to `out.js`)
- `node lex.js` (which runs `cjs-module-lexer` on `out.js`)

**Comment by guybedford** [comment]
Thanks for reporting. Per the policy on this project, the patterns supported were fully stabilized at the point this library started shipping, so that no new detections will be added. Otherwise that risks code that works in new Node.js versions mysteriously breaking in older versions and that would be very very difficult to debug.

I've posted a PR in https://github.com/nodejs/cjs-module-lexer/pull/80 to better explain this as part of the project status.

**Comment by yyx990803** [comment]
/cc @lukastaegert

If Node.js will no longer update the detection mechanism, maybe Rollup will have to adapt to this case? I think it comes down to the use of `for ... in` vs `Object.keys(...).forEach(...)`.

**Comment by guybedford** [comment]
Unfortunately that basically is the situation today - where build tools do have to accept these semantics as part of their targets now. It's ended up the lesser of the evils, and my apologies we couldn't do better!

**Comment by guybedford** [comment]
esbuild ended up just using a hinting pattern itself - https://github.com/nodejs/cjs-module-lexer/pull/46

**Comment by lukastaegert** [comment]
I think the change was done to allow some level of improved ES3 compatibility via this option. But I see that this is only affecting a very minor group of users and since `.forEach` can easily be polyfilled, I will revert it to the other pattern.

**Comment by lukastaegert** [comment]
Actually, looking at this longer, it was not about ES3 compatibility. I wanted to switch to the other pattern altogether because it is smaller and more performant, but it cannot be used for live bindings because `var k` is being reassigned. If you allow `const`, then it will also use a `for..in` loop with live-bindings: https://rollupjs.org/repl/?version=3.11.0&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQSUyMiUyMiUyQyUyMm1vZHVsZXMlMjIlM0ElNUIlN0IlMjJjb2RlJTIyJTNBJTIyZXhwb3J0JTIwKiUyMGZyb20lMjAnZm9vJyUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTJDJTIybmFtZSUyMiUzQSUyMm1haW4uanMlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyb3V0cHV0JTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyY2pzJTIyJTJDJTIyZ2VuZXJhdGVkQ29kZSUyMiUzQSU3QiUyMmNvbnN0QmluZGluZ3MlMjIlM0F0cnVlJTdEJTdEJTdEJTdE

But yes, I will revert this change.

**Comment by lukastaegert** [comment]
Fix at rollup/rollup#4826

**Comment by yyx990803** [comment]
Thanks @lukastaegert - I will close this one now.
